### PR TITLE
Fix quantized Conv1d same padding with even kernels in XNNPACK

### LINKED ISSUE #20558: Quantized Conv1d with even kernel + padding="same" produces incorrect XNNPACK output
### 🐛 Describe the bug

A statically-quantized nn.Conv1d with an even kernel_size and padding="same" lowers without error but produces numerically wrong results on the XNNPACK delegate. Odd kernels are fine. This is independent of the constant_pad_nd-fold work — I verified it reproduces with byte-identical error on the commit before any of that code.

```
import torch
from torch.export import export
from executorch.exir import to_edge_transform_and_lower
from executorch.backends.xnnpack.partition.xnnpack_partitioner import XnnpackPartitioner
from executorch.backends.xnnpack.quantizer.xnnpack_quantizer import (
    XNNPACKQuantizer, get_symmetric_quantization_config,
)
from executorch.extension.pybindings.portable_lib import _load_for_executorch_from_buffer
from torchao.quantization.pt2e.quantize_pt2e import convert_pt2e, prepare_pt2e

def run(kernel_size):
    class M(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.conv = torch.nn.Conv1d(2, 4, kernel_size=kernel_size, padding="same")
        def forward(self, x):
            return self.conv(x)

    ex = (torch.randn(1, 2, 16),)
    model = export(M().eval(), ex).module()
    q = XNNPACKQuantizer().set_global(get_symmetric_quantization_config())
    m = prepare_pt2e(model, q); m(*ex); m = convert_pt2e(m)
    prog = to_edge_transform_and_lower(export(m, ex), partitioner=[XnnpackPartitioner()]).to_executorch()
    et = _load_for_executorch_from_buffer(prog.buffer)
    diff = (et([*ex])[0] - m(*ex)).abs().max().item()
    print(f"kernel_size={kernel_size}: max abs diff = {diff:.5f}")

run(3)  # odd  -> 0.00000  (correct, quant noise only)
run(4)  # even -> ~1.4     (WRONG)
```

Observed: kernel_size=3 → 0.00000; kernel_size=4 → ~1.4 max abs diff vs the quantized reference.
Expected: even-kernel diff comparable to odd (small quant noise).

### Versions

Mainline

### PR DESCRIPTION
**Pull request by SakshamKapoor2911**
Fixes #20558.
Related to #20553.

## Summary

Quantized `nn.Conv1d(..., padding="same")` with an even kernel exports with
asymmetric padding (unequal left/right amounts), which cannot be folded into
the convolution's symmetric padding field. The original pad-folding approach
in this PR was replaced with the explicit-PAD approach from #20553:

- **InsertPadQDQPass** — inserts implicit quantize/dequantize pairs after
  constant_pad_nd nodes in quantized contexts so they serialize as quantized
  static pads. Refactored with additional guard checks (pad_value, pad_amounts,
  negative amounts) and correct idempotency.
- **ConvolutionConfig._get_act_deps** — pulls zero-valued constant_pad_nd nodes
  (and their QDQ chain if InsertPadQDQPass already ran) into the convolution's
  partition for both 1D and 2D convs. The merged method replaces separate 1D
  and 2D implementations that existed on this branch.
- **Regression test** — quantized Conv1d even-kernel same-padding covering both
  symmetric and asymmetric pad cases, validated by numerical comparison.
- Removed unrelated __init__.py re-ordering and no-op conv1d_unsqueeze_pass
  changes.

With these changes, an even-kernel `padding="same"` conv1d graph:

```
dequant -> constant_pad_nd -> convolution
```

becomes (after XNNPACK preprocessing and partitioning):

```
[dequant -> pad -> q -> dq -> conv]  (single XNNPACK delegate)
```

## Test plan

```
python -m pytest backends/xnnpack/test/ops/test_conv1d.py -q
# 7 passed (1 new regression test with 4 subTests)

python -m pytest backends/xnnpack/test/ops/test_conv2d.py -q
# 32 passed

python -m pytest backends/xnnpack/test/passes/test_insert_pad_qdq.py -q
# 3 passed

python -m pytest backends/xnnpack/test/ops/test_static_constant_pad.py -q
# 8 passed

lintrunner -a
# No lint issues
```

cc @GregoryComer @digantdesai @cbilgin @JakeStevens @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

File: backends/xnnpack/_passes/conv1d_unsqueeze_pass.py [hunk 1]
```diff
@@ -159,8 +159,42 @@ def call(self, graph_module: torch.fx.GraphModule):
 
                     # c. Add unsqueeze to input (3d -> 4d) and squeeze to output (4d -> 3d)
                     # unsqueeze -> conv2d -> squeeze
+                    input_node = node.args[0]
+                    if (
+                        isinstance(input_node, torch.fx.Node)
+                        and input_node.target
+                        == exir_ops.edge.aten.constant_pad_nd.default
+                        and len(input_node.users) == 1
+                    ):
```
[the remark this answers is not in the document]
Comment (SakshamKapoor2911) [reply backends/xnnpack/_passes/conv1d_unsqueeze_pass.py]: Good catch. Fixed in 529a79208 by gating both the Conv1d partition inclusion and fold logic on non-transposed convolutions (not node.args[6]), so explicit input padding is not folded into XNNPACK padding for transposed conv semantics.

File: backends/arm/_passes/arm_pass_utils.py [hunk 2]
```diff
@@ -40,8 +40,10 @@
 _Dim = int | torch.SymInt
 
 
-def is_submodule_node(node: torch.fx.Node):
-    if node.op not in ("get_attr", "placeholder"):
+def is_submodule_node(node: torch.fx.Node) -> bool:
```
Comment (JakeStevens) [inline backends/arm/_passes/arm_pass_utils.py]: Why are these changes in this PR?

File: backends/cortex_m/quantizer/pattern_matcher.py [hunk 3]
```diff
@@ -69,7 +69,11 @@ def _validate_match(
             return PatternMatchResult(match, False, self.REJECT_PREVIOUSLY_ANNOTATED)
 
         # Reject match if it contains a node that has an input which is too large to be quantized
-        if any(_is_large_scalar(node, node.graph.owning_module) for node in match):
+        if any(
```
Comment (JakeStevens) [inline backends/cortex_m/quantizer/pattern_matcher.py]: Same comment as above, if these changes are unrelated let's address in separate PRs

File: backends/arm/_passes/arm_pass_utils.py (hunk 2 above)
Comment (SakshamKapoor2911) [reply backends/arm/_passes/arm_pass_utils.py]: Apologies for the noise — these were opportunistic mypy type fixes I made while navigating the files during the Conv1d work. I've removed them from this branch. Happy to open a separate PR targeting backends/arm and backends/cortex_m if the type improvements are worth having upstream.

File: backends/cortex_m/quantizer/pattern_matcher.py (hunk 3 above)
Comment (SakshamKapoor2911) [reply backends/cortex_m/quantizer/pattern_matcher.py]: Same as above, I reversed these to focus on this PR/issue.

File: backends/arm/_passes/arm_pass_utils.py (hunk 2 above)
Comment (JakeStevens) [reply backends/arm/_passes/arm_pass_utils.py]: All good, feel free to open a PR for those, assign to me, and I can review that too!

**Comment by JakeStevens** [comment]
The PR for 2D has landed:

https://github.com/pytorch/executorch/pull/20553

You can take the same approach here; have an explicit PAD op, which gets executed in XNNPACK. We can then later fold that XNNPACK pad in the delegated graph, this is a follow up to both of these PRs.

As is, this code will also fail on:

```
class M(torch.nn.Module):
    def __init__(self):
          super().__init__()
          self.conv = torch.nn.Conv1d(1, 8, kernel_size=4, padding="same")
      def forward(self, x):
          out = torch.reshape(x, (1, 1, 256))
          out = self.conv(out)
          return torch.flatten(out, 1)   # even kernel + this line => crash
```

**Comment by SakshamKapoor2911** [comment]
Addressed review feedback: adopted the same approach as #20553 — keep the `constant_pad_nd` as an explicit PAD op executed in XNNPACK instead of folding it into asymmetric Conv2d padding.

Changes:
1. **Removed pad-folding** from `Conv1dUnsqueezePass` — the pad is no longer absorbed into asymmetric conv padding
2. **Removed `Conv1dFoldedPadMetaPass`** — no longer needed since tensor metadata propagates correctly with the standalone pad
3. **Reverted `xnnpack_input_padding`** handling in `op_conv2d.py` — symmetric padding from ATen args is used again
4. **Added `InsertPadQDQPass`** — inserts implicit QDQ after quantized-context pads so they serialize as quantized static pads (same mechanism as #20553)
5. The crash case with `conv1d(padding='same', even kernel) → flatten` is fixed because the pad is now a standalone op with correct shape propagation

The pad can be folded in the delegated graph later as a follow-up optimization (per the review suggestion).

**Comment by JakeStevens** [comment]
I believe you need to rebase and make sure the commit I shared is landed in your branch, there are merge conflicts from duplicated code

**Comment by SakshamKapoor2911** [comment]
Rebased onto current main and resolved the duplicated-code conflict. The conflict was an add/add in `insert_pad_qdq.py` — #20553 and this PR each introduced an `InsertPadQDQPass`. Kept #20553's canonical version (already landed on main) and dropped the duplicate import from `__init__.py`. No other files conflicted. PR is ready for another look.

**Comment by JakeStevens** [comment]
Please setup and run lintrunner to resolve the failures. See

https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#lintrunner

Also it seems like you've duplicated the get act dep in Gemm config instead of only updating as necessary.

I have not fully reviewed but I did notice that your comment said it kept the existing pad pass but the commit has substantial changes to it.

**Comment by SakshamKapoor2911** [comment]
Addressed the remaining review points:

**Lint**: `lintrunner -a` passes cleanly on all 5 changed files. No lint issues remain.

**Gemm config `_get_act_deps`**: The method extends the existing 2D-only pad absorption check from #20553 to also handle 1D convs and the QDQ-pair pattern (`dequant -> pad -> q -> dq -> conv`) that `InsertPadQDQPass` introduces. The fallthrough to `super()._get_act_deps()` is preserved at the end, so there is no duplication -- the method replaces the narrower 2D-only check with a version that covers both dimensions. Happy to refactor further if you have a specific structure in mind.

**`insert_pad_qdq.py`**: The changes are not purely additive because the pass needed to handle Conv1d's 1D constant_pad_nd and be idempotent (skip already-quantized pads). The core mechanism (insert QDQ after pad in a quantized context) is the same as the original.

**Comment by JakeStevens** [comment]
@claude please review

**Comment by SakshamKapoor2911** [comment]
Addressed all 3 items from Claude review:

**1. Per-tensor guard restored** (`insert_pad_qdq.py`)
Added `is_per_tensor(pad_input)` check alongside `is_dequant()`. This ensures only `dequantize_per_tensor.default` (per-tensor) activations get the implicit QDQ inserted — per-channel/per-token variants that produce different `q_params` shapes are excluded, matching the safety guard on `main`.

**2. Removed unreachable QDQ-chain branch** (`gemm_configs.py`)
Confirmed the `InsertPadQDQPass` runs during `XnnpackBackend.preprocess` (via `XNNPACKPassManager` at `_passes/__init__.py:85`), which is *after* partitioning. So at `_get_act_deps` call-graph time, the pattern is always `dequant -> pad -> conv` — the `dequant -> pad -> q -> dq -> conv` path is unreachable. Removed that branch.

**3. Minor cleanups**
- `modified` flag now accurately tracks whether any pad was actually modified
- Removed redundant `len(pad_amounts) % 2 == 0` (always true for `constant_pad_nd`)
- Removed duplicate `pad_value != 0.0` early return that duplicated the natural fallthrough

**Comment by SakshamKapoor2911** [comment]
@JakeStevens All 3 items from the Claude review above are addressed and pushed. Could you take another look?

File: backends/xnnpack/_passes/insert_pad_qdq.py [hunk 4]
```diff
@@ -4,81 +4,85 @@
...
+        for node in list(graph.nodes):
             if (
-                pad.op != "call_function"
-                or pad.target != exir_ops.edge.aten.constant_pad_nd.default
+                node.op != "call_function"
+                or node.target != exir_ops.edge.aten.constant_pad_nd.default
             ):
                 continue
 
-            # Only per-tensor static activations are handled: _insert_qdq_after
```
Comment (JakeStevens) [inline backends/xnnpack/_passes/insert_pad_qdq.py:63]: These comments were removed but what they were explaining was not fully followed.

In the original, we are allowing only _per_tensor.default, and then using that in the create_node. So, everything matched. This is no longer the case in the PR, see below

File: backends/xnnpack/_passes/insert_pad_qdq.py [hunk 5]
```diff
@@ -4,81 +4,85 @@
...
-            # dequant would supply mismatched args).
-            dq = pad.args[0]
-            if (
-                not isinstance(dq, torch.fx.Node)
-                or dq.target
-                != exir_ops.edge.quantized_decomposed.dequantize_per_tensor.default
+            pad_input = node.args[0]
+            if not (
+                isinstance(pad_input, torch.fx.Node)
+                and is_dequant(pad_input)
```
Comment (JakeStevens) [inline backends/xnnpack/_passes/insert_pad_qdq.py:46]: is_dequant now matches:

```
_DQ_OPS = {
    "dequantize_per_tensor.tensor",
    "dequantize_per_tensor.default",
    "dequantize_per_channel.default",
    "dequantize_per_channel_group.default",
    "dequantize_per_token.default",
    "dequantize_affine.default",
}
```

but then below, we only use the _per_tensor.default when rebuilding the graph.

**Review by JakeStevens — CHANGES_REQUESTED** [review]

**Comment by SakshamKapoor2911** [comment]
Addressed the review feedback: removed the redundant `is_dequant` check. `is_per_tensor` already verifies the node is a quant/dequant node with `per_tensor` in its target name, so the broader `is_dequant` (which now matches per_channel, per_token, etc.) was both redundant and confusing — it created a mismatch with the `create_node` targets below which always emit `per_tensor.default` variants.

```diff
- is_dequant(pad_input)
- and is_per_tensor(pad_input)
+ and is_per_tensor(pad_input)
```

@JakeStevens — ready for re-review when you get a chance.

**Comment by JakeStevens** [comment]
I'm sorry if I was not clear; the issue is not the redundant quant check, the issue is that before we checked on the *full* signature, and then replaced with that exact signature. Now, we match on both;
```
    "dequantize_per_tensor.tensor",
    "dequantize_per_tensor.default",
```

but insert only the .default version. 

So, you can have a _per_tensor.tensor quantized node and then we make the padding .default

Let us stick with just static/.default for now as that is what is covered by tests etc, we can land and if youd like follow up to see if .tensor can be supported too.

**Comment by SakshamKapoor2911** [comment]
@JakeStevens Feedback addressed -- tightened the dequant check in InsertPadQDQPass back to exact signature matching.

- Replaced is_per_tensor(pad_input) with is_dequant(pad_input) and pad_input.target == dequantize_per_tensor.default
- This matches only dequantize_per_tensor.default (not .tensor), so the inserted quantize_per_tensor.default node always receives matching scalar-typed q_params
- The extra is_dequant gate is a fast-path guard, keeping the intent readable

The .tensor variant can be supported in a follow-up if needed -- for now, only the covered/tested .default path is matched. Ready for re-review when you get a chance.

**Review by JakeStevens — APPROVED** [review]

**Comment by JakeStevens** [comment]
@SakshamKapoor2911  thank you for working through this!
